File: src/gbml/solution/michigan_solution.py
import copy
import time

import numpy as np

from src.fuzzy.knowledge.knowledge import Knowledge
from src.gbml.solution.abstract_solution import AbstractSolution


class MichiganSolution(AbstractSolution):
    _bounds = None
    _rule = None
    _rule_builder = None

    def __init__(self, num_objectives, num_constraints, rule_builder, bounds=None, michigan_solution=None):
        if bounds is None:
            bounds = MichiganSolution.make_bounds(knowledge=rule_builder.get_knowledge())

        self._bounds = bounds
        self._rule_builder = rule_builder

        super().__init__(len(bounds), num_objectives, num_constraints)

        if michigan_solution is not None:
            self._rule = copy.copy(michigan_solution.get_rule())
            self.set_vars(self._rule.get_antecedent().get_antecedent_indices())
            return

        cnt = 0
        is_rejected = True
        while is_rejected:
            cnt += 1
            self.create_rule()
            is_rejected = self._rule.is_rejected_class_label()
            if cnt > 1000:
                raise Exception("Exceeded maximum number of trials to generate rule")

    @staticmethod
    def make_bounds(knowledge):
        num_dim = knowledge.get_num_dim()

        return np.array([(0, knowledge.get_num_fuzzy_sets(dim_i)-1) for dim_i in range(num_dim)], dtype=object)

    def get_lower_bound(self, index):
        return self._bounds[index][0]

    def get_upper_bound(self, index):
        return self._bounds[index][1]

    def set_vars(self, new_vars):
        self._vars = new_vars

    def create_rule(self, michigan_solution=None):
        if michigan_solution is None:
            antecedent_indices = self._rule_builder.create_antecedent_indices()
            self.set_vars(antecedent_indices)
        else:
            raise Exception("Not yet implemented")
            # self.set_vars(self._rule_builder.create_antecedent_indices(michigan_solution))  # TODO: Not yet implemented
        self.learning()

    def learning(self):
        if self._vars is None:
            raise Exception("Vars is not defined")

        if self._rule is None or self._rule.get_antecedent() is None:
            antecedent_object = self._rule_builder.create_antecedent_from_indices(self._vars)
            self._rule = self._rule_builder.create(antecedent_object)
        else:
            antecedent_object = self._rule.get_antecedent()
            antecedent_object.set_antecedent_indices(self._vars)
            self._rule.set_consequent(self._rule_builder.create_consequent(antecedent_object))

    def get_fitness_value(self, in_vector):
        return self._rule.get_fitness_value(in_vector)

    def get_rule_length(self):
        return self._rule.get_rule_length(self.get_vars_array())

    def get_class_label(self):
        return self._rule.get_class_label()

    def get_rule_weight(self):
        return self._rule.get_rule_weight()

    def get_vars_array(self):
        return np.copy(self._vars)

    def get_rule(self):
        return self._rule

    def get_rule_builder(self):
        return self._rule_builder

    def get_consequent(self):
        return self._rule.get_consequent()

    def get_antecedent(self):
        return self._rule.get_antecedent()

    def get_compatible_grade(self, attribute_vector):
        return self._rule.get_compatible_grade(attribute_vector)

    def get_compatible_grade_value(self, attribute_vector):
        return self._rule.get_compatible_grade_value(attribute_vector)

    def __copy__(self):
        return MichiganSolution(self.get_num_objectives(), self.get_num_constraints(), copy.copy(self._rule_builder), michigan_solution=self)

    class MichiganSolutionBuilder(AbstractSolution.SolutionBuilderCore):
        def __init__(self, bounds, num_objectives, num_constraints, rule_builder):
            super().__init__(bounds, num_objectives, num_constraints, rule_builder)

        def create_michigan_solution(self, num_solutions=1):
            start = time.time()
            solutions = []
            bounds = self._bounds
            if bounds is None:
                bounds = MichiganSolution.make_bounds(self._rule_builder.get_knowledge())

            for i in range(num_solutions):
                solutions.append(MichiganSolution(self._num_objectives, self._num_constraints, self._rule_builder, bounds))
                # solutions[i].set_attribute(attribute_id, 0) # TODO: check usage in java version
                # solutions[i].set_attribute(attribute_id_fitness, 0)

            elapsed = time.time() - start
            # print(elapsed) # approx 0.1

            return solutions

        def __copy__(self):
            return MichiganSolution.MichiganSolutionBuilder(self._bounds, self._num_objectives, self._num_constraints, copy.copy(self._rule_builder))

File: src/mofgbmlpy/core/gbml/solution/michigan_solution.cpp
#include "michigan_solution.hpp"

#include <sstream>
#include <utility>

MichiganSolution::MichiganSolution(std::shared_ptr<RandomGenerator> random_gen, int num_objectives, int num_constraints,
    std::shared_ptr<RuleBuilderCore> rule_builder, Pattern& pattern, bool do_init_vars): AbstractSolution(num_objectives, num_constraints), num_wins(0), fitness(0), rule_builder(std::move(rule_builder)), random_gen(std::move(random_gen)), are_scores_updated(false)
{
    if (do_init_vars) {
        create_rule(pattern);
        bool is_rejected = rule->is_rejected_class_label();
        int cnt = 0;
        std::shared_ptr<Dataset> training_data_set = rule_builder->get_training_dataset();

        while (is_rejected) {
            if (cnt > 1000) {
                throw std::runtime_error("Exceeded maximum number of trials to generate rule");
            }

            cnt ++;
            int pattern_idx = random_gen->randint(0, training_data_set->get_size());
            Pattern* pattern = training_data_set->get_pattern(pattern_idx);
            create_rule(*pattern);
            is_rejected = rule->is_rejected_class_label();
        }
    }
}

MichiganSolution::MichiganSolution(std::shared_ptr<RandomGenerator> random_gen, int num_objectives, int num_constraints,
    std::shared_ptr<RuleBuilderCore> rule_builder, bool do_init_vars): AbstractSolution(num_objectives, num_constraints), num_wins(0), fitness(0), rule_builder(std::move(rule_builder)), random_gen(std::move(random_gen)), are_scores_updated(false)
{
    if (do_init_vars) {
        int cnt = 0;
        bool is_rejected = true;
        while (is_rejected) {
            cnt++;
            create_rule();
            is_rejected = rule->is_rejected_class_label();
            if (cnt > 1000) {
                throw std::runtime_error("Exceeded maximum number of trials to generate rule");
            }
        }
    }
}

MichiganSolution::MichiganSolution(const MichiganSolution& other): AbstractSolution(other.get_num_objectives(), other.get_num_constraints()), num_wins(other.num_wins), fitness(other.fitness), rule_builder(other.rule_builder->clone()), random_gen(other.random_gen), are_scores_updated(other.are_scores_updated), rule(other.rule->clone()), vars(other.vars) {}

void MichiganSolution::set_scores_update_status(bool new_status)
{
    are_scores_updated = new_status;
}

void MichiganSolution::create_rule(const Pattern& pattern)
{
    set_vars(rule_builder->create_antecedent_indices(pattern)[0]);
    learning();
}

void MichiganSolution::create_rule()
{
    set_vars(rule_builder->create_antecedent_indices(1)[0]);
    learning();
}

void MichiganSolution::learning(const Dataset& dataset)
{
    if (rule == nullptr || rule->get_antecedent() == nullptr) {
        Antecedent* antecedent_object = rule_builder->create_antecedent_from_indices(vars);
        rule = std::shared_ptr<AbstractRule>(rule_builder->create(*antecedent_object));
    } else {
        std::shared_ptr<Antecedent> antecedent_object = rule->get_antecedent();
        antecedent_object->set_antecedent_indices(vars);
        rule->set_consequent(rule_builder->create_consequent(*antecedent_object, dataset));
    }
}

void MichiganSolution::learning()
{
    if (rule == nullptr || rule->get_antecedent() == nullptr) {
        Antecedent* antecedent_object = rule_builder->create_antecedent_from_indices(vars);
        rule = std::shared_ptr<AbstractRule>(rule_builder->create(*antecedent_object));
    } else {
        std::shared_ptr<Antecedent> antecedent_object = rule->get_antecedent();
        antecedent_object->set_antecedent_indices(vars);
        rule->set_consequent(rule_builder->create_consequent(*antecedent_object));
    }
}

double MichiganSolution::get_fitness_value(const std::vector<double>& in_vector) const
{
    return rule->get_fitness_value(in_vector);
}

int MichiganSolution::get_length() const
{
    return rule->get_length();
}

std::shared_ptr<AbstractClassLabel> MichiganSolution::get_class_label() const
{
    return rule->get_class_label();
}

std::shared_ptr<AbstractRuleWeight> MichiganSolution::get_rule_weight() const
{
    return rule->get_rule_weight();
}

std::shared_ptr<AbstractRule> MichiganSolution::get_rule() const
{
    return rule;
}

std::shared_ptr<RuleBuilderCore> MichiganSolution::get_rule_builder() const
{
    return rule_builder;
}

std::shared_ptr<AbstractConsequent> MichiganSolution::get_consequent() const
{
    return rule->get_consequent();
}

std::shared_ptr<Antecedent> MichiganSolution::get_antecedent() const
{
    return rule->get_antecedent();
}

std::vector<double> MichiganSolution::get_membership_values(const std::vector<double>& attribute_vector) const
{
    return rule->get_membership_values(attribute_vector);
}

double MichiganSolution::get_compatible_grade_value(const std::vector<double>& attribute_vector) const
{
    return rule->get_compatible_grade_value(attribute_vector);
}

void MichiganSolution::reset_num_wins()
{
    num_wins = 0;
    are_scores_updated = false;
}

void MichiganSolution::reset_fitness()
{
    num_wins = 0;
    are_scores_updated = false;
}

void MichiganSolution::inc_num_wins()
{
    num_wins++;
}

void MichiganSolution::inc_fitness()
{
    fitness++;
}

int MichiganSolution::get_num_wins() const
{
    if (!are_scores_updated) {
        throw std::runtime_error("Scores are not updated. Call learning() before accessing scores.");
    }
    return num_wins;
}

int MichiganSolution::get_fitness() const
{
    if (!are_scores_updated) {
        throw std::runtime_error("Scores are not updated. Call learning() before accessing scores.");
    }
    return fitness;
}

void MichiganSolution::clear_vars()
{
    vars = std::vector<int>();
    are_scores_updated = false;
}

std::vector<int> MichiganSolution::get_vars() const
{
    return vars;
}

int MichiganSolution::get_var(int index) const
{
    return vars[index];
}

void MichiganSolution::set_var(int index, int value)
{
    vars[index] = value;
    are_scores_updated = false;
}

void MichiganSolution::set_vars(const std::vector<int>& new_vars)
{
    vars = new_vars;
    are_scores_updated = false;
}

int MichiganSolution::get_num_vars() const
{
    return vars.size();
}

void MichiganSolution::set_antecedent_knowledge(Knowledge& knowledge)
{
    get_antecedent()->set_knowledge(&knowledge);
}

bool MichiganSolution::is_rejected()
{
    return rule->is_rejected_class_label();
}

MichiganSolution::operator std::string() const
{
    std::ostringstream oss;
    oss << "(MichiganSolution) Variables: [";

    for (int i = 0; i < get_num_vars(); ++i) {
        oss << vars[i] << " ";
    }

    oss << "], Rule weight: " << (rule ? std::string(*rule->get_rule_weight()) : "null")
        << ", Class label: " << (rule ? std::string(*rule->get_class_label()) : "null");

    oss << "], Objectives: [";
    for (int i = 0; i < get_num_objectives(); ++i) {
        oss << get_objective(i) << " ";
    }

    oss << "], Attributes: {Number of classifier patterns: " << fitness
        << ", Number of wins: " << num_wins;
    for (const auto& [key, val] : attributes) {
        oss << ", " << key << ": ";
        if (val.type() == typeid(std::string)) {
            oss << std::any_cast<std::string>(val);
        } else if (val.type() == typeid(int)) {
            oss << std::any_cast<int>(val);
        } else if (val.type() == typeid(double)) {
            oss << std::any_cast<double>(val);
        } else if (val.type() == typeid(float)) {
            oss << std::any_cast<float>(val);
        } else {
            oss << "<unknown type>";
        }
    }
    oss << "}";
    return oss.str();
}

bool MichiganSolution::operator==(const AbstractSolution& other) const
{
    MichiganSolution* other_michigan = dynamic_cast<MichiganSolution*>(&const_cast<AbstractSolution&>(other));
    if (other_michigan == nullptr) {
        return false;
    }
    if(vars != other_michigan->vars) {
        return false;
    }
    return true;
}

MichiganSolution* MichiganSolution::clone() const
{
    return new MichiganSolution(*this);
}
